src/llvm-alloc-opt.cpp

```diff
@@ -0,0 +1,746 @@
+    T_ppjlvalue_der = PointerType::get(T_prjlvalue, AddressSpace::Derived);
+
+    T_int8 = Type::getInt8Ty(*ctx);
+    T_int32 = Type::getInt32Ty(*ctx);
+    T_int64 = Type::getInt64Ty(*ctx);
+    T_size = sizeof(void*) == 8 ? T_int64 : T_int32;
+    T_pint8 = PointerType::get(T_int8, 0);
+
+    if (!(pool_alloc = M.getFunction("jl_gc_pool_alloc"))) {
+        std::vector<Type*> alloc_pool_args(0);
+        alloc_pool_args.push_back(T_pint8);
+        alloc_pool_args.push_back(T_int32);
+        alloc_pool_args.push_back(T_int32);
+        pool_alloc = Function::Create(FunctionType::get(T_prjlvalue, alloc_pool_args, false),
+                                      Function::ExternalLinkage, "jl_gc_pool_alloc", &M);
+        addRetNoAlias(pool_alloc);
+    }
+    if (!(big_alloc = M.getFunction("jl_gc_big_alloc"))) {
+        std::vector<Type*> alloc_big_args(0);
+        alloc_big_args.push_back(T_pint8);
+        alloc_big_args.push_back(T_size);
+        big_alloc = Function::Create(FunctionType::get(T_prjlvalue, alloc_big_args, false),
+                                     Function::ExternalLinkage, "jl_gc_big_alloc", &M);
+        addRetNoAlias(big_alloc);
+    }
+
+#if JL_LLVM_VERSION >= 50000
+    lifetime_start = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_start, { T_pint8 });
+    lifetime_end = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_end, { T_pint8 });
+#else
+    lifetime_start = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_start);
+    lifetime_end = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_end);
+#endif
+
+    MDNode *tbaa_data;
+    MDNode *tbaa_data_scalar;
+    std::tie(tbaa_data, tbaa_data_scalar) = tbaa_make_child("jtbaa_data");
+    tbaa_tag = tbaa_make_child("jtbaa_tag", tbaa_data_scalar).first;
+
+    return true;
+}
+
+bool AllocOpt::checkInst(Instruction *I, CheckInstStack &stack, std::set<Instruction*> &uses,
+                         bool &ignore_tag)
+{
+    uses.clear();
+    if (I->user_empty())
+        return true;
+    CheckInstFrame cur{I, 0, I->user_begin(), I->user_end()};
+    stack.clear();
+
+    // Recursion
+    auto push_inst = [&] (Instruction *inst) {
+        if (cur.user_it != cur.user_end)
+            stack.push_back(cur);
+        cur.parent = inst;
+        cur.user_it = inst->user_begin();
+        cur.user_end = inst->user_end();
+    };
+
+    auto check_inst = [&] (Instruction *inst) {
+        if (isa<LoadInst>(inst))
+            return true;
+        if (auto call = dyn_cast<CallInst>(inst)) {
+            // TODO: on LLVM 5.0 we may need to handle certain llvm intrinsics
+            // including `memcpy`, `memset` etc. We might also need to handle
+            // `memcmp` by coverting to our own intrinsic and lower it after the gc root pass.
+            if (ptr_from_objref && ptr_from_objref == call->getCalledFunction())
+                return true;
+            // Only use in argument counts, use in operand bundle doesn't since it cannot escape.
```

Comment:
Only skip the "jl_roots" arg bundle? In general arg bundles escape in unknown ways per the spec.